server/src/main/java/org/elasticsearch/cluster/metadata/Template.java

```diff
@@ -345,8 +397,123 @@ public Builder lifecycle(DataStreamLifecycle.Builder lifecycle) {
             return this;
         }
 
+        public Builder dataStreamOptions(DataStreamOptions.Template dataStreamOptions) {
+            this.dataStreamOptions = ExplicitlyNullable.create(dataStreamOptions);
+            return this;
+        }
+
+        public Builder dataStreamOptions(ExplicitlyNullable<DataStreamOptions.Template> dataStreamOptions) {
+            this.dataStreamOptions = dataStreamOptions;
+            return this;
+        }
+
         public Template build() {
-            return new Template(settings, mappings, aliases, lifecycle);
+            return new Template(settings, mappings, aliases, lifecycle, dataStreamOptions);
+        }
+    }
+
+    public static class ExplicitlyNullable<T> {
```

Comment:
Given: This object captures three states by doing the following: 

1) Explicitly Set - Exists and has a value
2) Explicitly Unset - Exists and has a null
3) Unspecified - Does not exist

(These state names are just descriptive. I'm not suggesting we formalize their names or anything).

After having looked through the PR a bit, I notice that by accepting a null pointer to this object as state 3 above we end up with `if property == null || property.isNull()` or similar in places.

I think we should do one of two things with this class:

1) Replace uses of it with `Optional<T>`
    * I don't think this class is adding much over the existing `Optional` aside from having a prescriptive name and avoiding IDE warnings.
2) Capture state 3 above inside of the class and make it `@NotNull` on `Template`, `DataStreamOptions`, and `DataStreamFailureStore`
    * Differentiates it from `Optional` more.
    * When we parse XContent, we can parse a value to state 1, explicit null to state 2, and at the point that we call the constructor in the parser we convert a null result there to be state 3.
    * Avoids any possibility of `if property == null || property.isNull()` code. 
    * Lets us encapsulate the states on the object itself instead of expecting callers to make sense of the special 3rd state of the whole thing being null.

Reply:
Interesting, I will work on it.

Reply:
@jbaiera I pushed the change in https://github.com/elastic/elasticsearch/pull/117357/commits/f1a1797ecef446b0851bd477189224397badcbb9. I try to abstract it a bit from the concept of null and focus on how we want it to behave. Not sure if it is better, but I see it  now as a way to unset a configuration. Curious about your thoughts :)

Reply:
I like the change to `ResettableValue`. It looks really good! My only thought is maybe we should change the "unset" state to be named something like "reset", since "unset" and "undefined" seem really close in their meaning in my head.